### PR TITLE
pebbleiter: implement ValueAndErr

### PR DESCRIPTION
**Pull request by jbowens**
Implement ValueAndErr on the storage/pebbleiter package's assertionIter to ensure correct use of value slices. Additionally, assert that no callers make use of the deprecated Value method.

Epic: none
Release note: none

### CODE REVIEW

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134466#-:-OB3UCzR-pnwDeOa9Dzh:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/storage/pebbleiter/crdb_test_on.go` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134466#-OB3SyN-CiwZTZD9tXPd:-OB3SyN-CiwZTZD9tXPe:b-v9zbu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1eebc1af216edcba89618524d815c068fe2c8a/pkg/storage/pebbleiter/crdb_test_on.go#L6)):*
> ```Go
> // included in the /LICENSE file.
> 
> //go:build crdb_test && !crdb_test_off
> ```

I don't think this actually works.. We don't build with these tags with `bazel`, we use tweaks in the `BUILD.bazel` files like here:
https://github.com/cockroachdb/cockroach/blob/c6485194a0f134cf4a1ca6d84193eab491111168/pkg/util/buildutil/BUILD.bazel#L10

I will work on improving this (the first step is #4117)
___
*[`pkg/storage/pebbleiter/crdb_test_on.go` line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134466#-OB3TjpK4zOb79aaNQhC:-OB3TjpK4zOb79aaNQhD:b-ub0n4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1eebc1af216edcba89618524d815c068fe2c8a/pkg/storage/pebbleiter/crdb_test_on.go#L131)):*
> ```Go
> 	idx := i.unsafeBufs.idx
> 	i.unsafeBufs.val[idx] = append(i.unsafeBufs.val[idx][:0], val...)
> 	// Preserve nil-ness to ensure we test exactly the behavior of Pebble.
> ```

[nit] This is a bit hard to follow, maybe
```
if err != nil {
  return nil, err
}
if len(val) == 0 {
  if val == nil {
    return nil, nil
  }
  return []byte{}, nil
}
```

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134466)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/storage/pebbleiter/crdb_test_on.go` line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134466#-OB3UGGm3iJPVCZYRVtD:-OB3UGGm3iJPVCZYRVtE:b-xjpt7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1eebc1af216edcba89618524d815c068fe2c8a/pkg/storage/pebbleiter/crdb_test_on.go#L124)):*
> ```Go
> }
> 
> func (i *assertionIter) ValueAndErr() ([]byte, error) {
> ```

We need this in the `crdb_test_off.go` too, no?

**Comment by RaduBerinde** [comment]
*[`pkg/storage/pebbleiter/crdb_test_on.go` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134466#-OB3SyN-CiwZTZD9tXPd:-OB3WgSb-foRro0edOsZ:b-l2jgr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1eebc1af216edcba89618524d815c068fe2c8a/pkg/storage/pebbleiter/crdb_test_on.go#L6)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think this actually works.. We don't build with these tags with `bazel`, we use tweaks in the `BUILD.bazel` files like here:
https://github.com/cockroachdb/cockroach/blob/c6485194a0f134cf4a1ca6d84193eab491111168/pkg/util/buildutil/BUILD.bazel#L10

I will work on improving this (the first step is #4117)
</blockquote></details>

Oh, never mind, this is in the cockroach repo and the corresponding `BUILD.bazel` does do this properly, sorry about that.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134466#-:-OB3WxI16-UITS5V-_ko:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/storage/pebbleiter/crdb_test_on.go` line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134466#-OB3UGGm3iJPVCZYRVtD:-OB3WtAG4_EBCqyq_1bG:b-tvhl98) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1eebc1af216edcba89618524d815c068fe2c8a/pkg/storage/pebbleiter/crdb_test_on.go#L124)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We need this in the `crdb_test_off.go` too, no?
</blockquote></details>

Never mind :)

**Comment by jbowens** [comment]
TFTR!

bors r=RaduBerinde
